File: Applied Course/4.Problem Solving/3.Problems on Linked List/3.Remove Nth Node From End of List.py
# https://leetcode.com/problems/remove-nth-node-from-end-of-list
# Logic-1 Counting
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy=ListNode(0)
        dummy.next=head
        #find length of list
        length=0
        first=head
        while first:
            length+=1
            first=first.next
        length-=n
        first=dummy
        while length>0:
            length-=1
            first=first.next
        # update the link
        first.next=first.next.next
        return dummy.next
# Logic-2 Two pointer Approach
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy=ListNode(0)
        dummy.next=head
        first=dummy
        second=dummy
        for i in range(n+1):
            first=first.next
        while first:
            first=first.next
            second=second.next
        second.next=second.next.next
        return dummy.next

File: Applied Course/4.Problem Solving/3.Problems on Linked List/3.Remove Nth Node From End of List.cpp
#include<bits/stdc++.h>
using namespace std;
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
// Logic-1 Counting
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int length=0;
        struct ListNode *dummy,*temp=head;
        dummy=(struct ListNode*)malloc(sizeof(struct ListNode));
        dummy->val=0;
        dummy->next=head;
        while(temp)
        {
            length++;
            temp=temp->next;
        }
        length-=n;
        temp=dummy;
        while(length--)
            temp=temp->next;
        temp->next=temp->next->next;
        return dummy->next;
    }
};
// Logic-2 Two Pointer Approach
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        struct ListNode *dummy;
        dummy=(struct ListNode*)malloc(sizeof(struct ListNode));
        dummy->next=head;
        struct ListNode *first=dummy,*second=dummy;
        for(int i=0;i<=n;i++)
            first=first->next;
        while(first)
        {
            first=first->next;
            second=second->next;
        }
        second->next=second->next->next;
        return dummy->next;
    }
};
int main()
{
    // https://leetcode.com/problems/remove-nth-node-from-end-of-list/
    return 0;
}